### PR TITLE
Use Result instead of GetAwaiter().GetResult when IsCompletetSuccessfully is true

### PR DESCRIPTION
**Pull request by Kahbazi**
Same as #13614

### CODE REVIEW

File: src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public async Task Invoke(HttpContext context)
             // Make sure we only ever call GetResult once on the TryEnterAsync ValueTask b/c it resets.
             bool result;
 
-            if (waitInQueueTask.IsCompleted)
+            if (waitInQueueTask.IsCompletedSuccessfully)
```
Comment (Tratcher) [inline src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs]: No, this skips the error condition.

Comment (Kahbazi) [reply src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs]: I was aware of that, I was thinking that this way if an error occurs it would raise here https://github.com/Kahbazi/AspNetCore/blob/ce97530633ba7d2a068e813b7bc2d3451487f0c2/src/Middleware/ConcurrencyLimiter/src/ConcurrencyLimiterMiddleware.cs#L63 and we have a better exception since await is used instead of Result.

**Comment by davidfowl** [comment]
Why is this change better?

**Comment by Kahbazi** [comment]
@davidfowl I run a benchmark and saw when `ValueTask` is completed successfully returning via Result if far better than `GetAwaiter().GetResult()`

**Comment by davidfowl** [comment]
> @davidfowl I run a benchmark and saw when ValueTask is completed successfully returning via Result if far better than GetAwaiter().GetResult()

Where's the benchmark? Can you show how it affects the middleware?

**Comment by Kahbazi** [comment]
Oh I thought you were talking about the `ValueTaskExtensions`.
The change on the middleware is just to avoid AggregateException in case of exception.
These two has nothing to do with each other. Maybe I should have done these in two separate PRs.

File: src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ public static Task GetAsTask(this in ValueTask<FlushResult> valueTask)
             if (valueTask.IsCompletedSuccessfully)
             {
                 // Signal consumption to the IValueTaskSource
-                valueTask.GetAwaiter().GetResult();
+                _ = valueTask.Result;
```
Comment (davidfowl) [inline src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs]: Revert this change.

Comment (Kahbazi) [reply src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs]: This was my main point in this PR. I thought using `Result` is better than `GetAwaiter().GetResult()` when `IsCompletedSuccessfully` is `true`. Could you please explain why this change is not good?

Also I found couple of similar scenario which are using `Result`
https://github.com/aspnet/AspNetCore/blob/b23ea5b6683b08e8d168ccf49e0fc8515077ee2e/src/Mvc/Mvc.Core/src/Infrastructure/ControllerActionInvoker.cs#L431
https://github.com/aspnet/AspNetCore/blob/5e575a3e64254932c1fd4937041a4e7426afcde4/src/Http/Routing/src/EndpointRoutingMiddleware.cs#L65
https://github.com/aspnet/AspNetCore/blob/0a1d68b8f375eb979c5722b8bedd57496749b1f8/src/Mvc/Mvc.RazorPages/src/Infrastructure/PageLoaderMatcherPolicy.cs#L79
https://github.com/aspnet/AspNetCore/blob/caa910ceeba5f2b2c02c47a23ead0ca31caea6f0/src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/TimingPipeFlusher.cs#L60
https://github.com/aspnet/AspNetCore/blob/94edcbf65bc68aeebc97be8347577304059a01e0/src/Mvc/Mvc.ViewFeatures/src/Controller.cs#L369

Is there any difference between theses and this case?

Comment (Kahbazi) [reply src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs]: /cc @pranavkm

**Comment by anurse** [comment]
I'm not sure I understand the purpose of this change now. Does using `Result` change the behavior at all? I know we use it in other places but we also [use `GetAwaiter().GetResult()` in plenty of places too](https://github.com/aspnet/AspNetCore/search?utf8=%E2%9C%93&q=GetAwaiter%28%29&type=) (i.e. we're already inconsistent :))

**Comment by Kahbazi** [comment]
It doesn't change the behavior, at least in this scenario when the task is completed. Also it saves creation of `ValueTaskAwaiter` struct.

**Comment by anurse** [comment]
> Also it saves creation of `ValueTaskAwaiter` struct.

Since it's a struct, that's a pretty minimal impact. Is there a measurable performance change?

**Comment by Kahbazi** [comment]
```
[Benchmark]
public bool Result()
{
    ValueTask<bool> task = new ValueTask<bool>(true);
    return task.Result;
}

[Benchmark]
public bool GetAwaiterGetResult()
{
    ValueTask<bool> task = new ValueTask<bool>(true);
    return task.GetAwaiter().GetResult();
}
```


Method | Mean | Error | StdDev | Median | Rank
-- | -- | -- | -- | -- | --
Result | 0.0129 ns | 0.0202 ns | 0.0353 ns | 0.0000 ns | 1
GetAwaiterGetResult | 10.6799 ns | 0.2601 ns | 0.2672 ns | 10.7538 ns | 2

**Comment by anurse** [comment]
A super micro benchmark like that isn't terribly useful though since it doesn't really show it in a proper context. A delta of 10ns is extremely small and wouldn't really matter in most scenarios.

I realize this is a bit nitpicky given the tiny size of the change but in general we err on the side of leaving existing code as-is for a scenario like this. The two code paths (`.Result` and `.GetAwaiter().GetResult()`) **are** different, even if they seem to have the same result.

I notice the previous PR linked (https://github.com/aspnet/AspNetCore/pull/13614) was also closed as we intended to use `GetAwaiter().GetResult()` there. I'm think we'd have the same position on this change. If there's a clear benefit to the product in changing to `Result` and we understand the costs and changes then we can consider something like this, but it feels like we don't have that here.

**Comment by Kahbazi** [comment]
I totally understand your point. The difference in performance is very little.
My other goal was to make this code similar to the other parts, that are using `Result` when there's a check on `IsCompleted`.

>  we err on the side of leaving existing code as-is for a scenario like this


Just a question, What about for future codes? which one should be used to scenario like this?

Also #13614 was closed at first but then it got reopened and got merged, so I kept looking for the similar scenarios 😁

**Comment by anurse** [comment]
> Also #13614 was closed at first but then it got reopened and got merged, so I kept looking for the similar scenarios 😁

Ah you're right! I mis-read the status :)

> Just a question, What about for future codes? which one should be used to scenario like this?

Honestly, I couldn't say. In reality, I think you are correct that the behavior is the same (or similar enough) in either case. We don't prescribe exactly what to do in every scenario, it's up to the individual contributor (internal or external) and code reviewers to decide on what is appropriate.

**Comment by Kahbazi** [comment]
Fair enough. I'm ok if you want to close this PR.
